openssl: split the (mostly empty) runtime dependencies of static builds into a separate output

monacoremo opened the pull request:
###### Motivation for this change

This closure size of all packages that statically link `openssl` is unnecessarily large, as they get a runtime dependency on the static openssl build.

This is because the paths to several mostly empty directories and files in `--openssldir` and `ENGINESDIR` that are being baked  into the `libcrypto.a` file: 

```bash
$ strings /nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/lib/libcrypto.a | grep /nix/store
/nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/etc/ssl/ct_log_list.cnf
OPENSSLDIR: "/nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/etc/ssl"
ENGINESDIR: "/nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/lib/engines-1.1"
/nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/lib/engines-1.1
/nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/etc/ssl/private
/nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/etc/ssl
/nix/store/qlw2d1jwjijm1g70f0p4cbwqf9ccmw78-openssl-1.1.1d/etc/ssl/certs
```

This is an attempt to reduce the runtime dependencies of packages that statically link openssl by:

* Moving the '--openssldir' to a separate output.

This is relatively simple and clean, but unfortunately not enough, as the ENGINESDIR, which is also being baked into the static library and it cannot be configured (hardcoded to be under `libdir` by openssl). So we also need to:

* Patch the ENGINESDIR to be in the separate output instead of `libdir` on static builds.

This is a bit messy, but I could't figure out a cleaner solution so far... The substitution is only done on static builds, as the dynamic `*.so` files in ENGINESDIR contain references to `$out` and cannot be moved to a separate output.

I would be very grateful for any hints on how to improve this PR!

In my use case, this reduces the closure size of e.g. https://github.com/PostgREST/postgrest by over 60%, from 13mb compressed to about 5mb.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date (added comments)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
openssl: split runtime dependencies of static builds into a separate output

ajs124 commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -34,9 +34,20 @@ let
       substituteInPlace crypto/async/arch/async_posix.h \
         --replace '!defined(__ANDROID__) && !defined(__OpenBSD__)' \
                   '!defined(__ANDROID__) && !defined(__OpenBSD__) && 0'
+    '' + optionalString (versionAtLeast version "1.1.1" && static)
```
Why does this depend on 1.1.1?

monacoremo replied:
It should probably also be good for 1.0.2. I'll test that and remove the check if it works!

monacoremo commented:
@ajs124 I've tested it with 1.0.2, the version check was not needed and I removed it.

I also renamed the new output from `openssldir` to `etc`, as it also contains the ENGINESDIR and reflects the content better, I think.

ajs124 commented:
This isn't a no-op for non-static builds, because of the changes in `postInstall` and `postFixup`.

It's probably still not a bad idea, though.

SuperSandro2000 reviewed (approved):
I like the idea

monacoremo commented:
I rebased on staging to resolve the merge conflict (rebased on `master` first by mistake, sorry for the automated review requests :-| ) and tested that `pkgs.openssl` and `pkgsMusl,openssl` evaluate and build on NixOS/x64.

SuperSandro2000 commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -134,6 +161,9 @@ let
       # "#!/usr/bin/env perl"
     ''
       substituteInPlace $out/bin/c_rehash --replace ${buildPackages.perl}/bin/perl "/usr/bin/env perl"
+    '' + lib.optionalString (!static)
+    ''
+      openssldir=$out
     '' +
     ''
```
```suggestion
    '' + lib.optionalString (!static) ''
      openssldir=$out
    '' + ''
```

SuperSandro2000 commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -120,6 +143,10 @@ let
 
+      # 'etc' is a separate output on static builds only.
+      etc=$out
+    '' +
+    ''
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -48,9 +48,22 @@ let
     '';
 
-    outputs = [ "bin" "dev" "out" "man" ] ++ optional withDocs "doc";
+    outputs =
+      [ "bin" "dev" "out" "man" ]
```
```suggestion
    outputs = [ "bin" "dev" "out" "man" ]
```

monacoremo commented:
Rebased on latest `staging`

robx commented:
This patch continues to be maintained in PostgREST, a version against current nixpkgs is at https://github.com/PostgREST/postgrest/blob/692b23abbd7c88f2aaf7b3fac18bd52a0eb09d76/nix/patches/nixpkgs-openssl-split-runtime-dependencies-of-static-builds.patch. I'm a bit curious whether it's realistic to get that applied to nixpkgs?

veprbl commented:
Quite realistic. But this needs a rebase.

robx commented:
> Quite realistic. But this needs a rebase.

I created a rebased PR at #182444, since @monacoremo isn't around.

fpletz commented:
As #182444 was merged we can close this. Thanks!
